clang/lib/Sema/SPIRVBuiltins.td

```diff
@@ -840,6 +840,13 @@ foreach AS = [GlobalAS, LocalAS, PrivateAS] in {
   def : SPVBuiltin<"GenericCastToPtrExplicit", [PointerType<Char, AS>, PointerType<Char, GenericAS>], Attr.Const>;
 }
 
+def : SPVBuiltin<"GenericCastToPtrExplicit_ToGlobal", [PointerType<Void, GlobalAS>, PointerType<ConstType<Void>, DefaultAS>, Int], Attr.Const>;
+def : SPVBuiltin<"GenericCastToPtrExplicit_ToGlobal", [PointerType<VolatileType<Void>, GlobalAS>, PointerType<VolatileType<ConstType<Void>>, DefaultAS>, Int], Attr.Const>;
+def : SPVBuiltin<"GenericCastToPtrExplicit_ToLocal", [PointerType<Void, LocalAS>, PointerType<ConstType<Void>, DefaultAS>, Int], Attr.Const>;
+def : SPVBuiltin<"GenericCastToPtrExplicit_ToLocal", [PointerType<VolatileType<Void>, LocalAS>, PointerType<VolatileType<ConstType<Void>>, DefaultAS>, Int], Attr.Const>;
+def : SPVBuiltin<"GenericCastToPtrExplicit_ToPrivate", [PointerType<Void, PrivateAS>, PointerType<ConstType<Void>, DefaultAS>, Int], Attr.Const>;
+def : SPVBuiltin<"GenericCastToPtrExplicit_ToPrivate", [PointerType<VolatileType<Void>, PrivateAS>, PointerType<VolatileType<ConstType<Void>>, DefaultAS>, Int], Attr.Const>;
```

Comment:
You are dropping const here. Only the address space should change here (https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpGenericCastToPtrExplicit).

```suggestion
foreach Ty = [Void, ConstType<Void>, VolatileType<Void>, VolatileType<ConstType<Void>>] in {
  def : SPVBuiltin<"GenericCastToPtrExplicit_ToGlobal", [PointerType<Ty, GlobalAS>, PointerType<Ty, DefaultAS>, Int], Attr.Const>;
  def : SPVBuiltin<"GenericCastToPtrExplicit_ToLocal", [PointerType<Ty, LocalAS>, PointerType<Ty, DefaultAS>, Int], Attr.Const>;
  def : SPVBuiltin<"GenericCastToPtrExplicit_ToPrivate", [PointerType<Ty, PrivateAS>, PointerType<Ty, DefaultAS>, Int], Attr.Const>;
}
```

Reply:
It is a bit silly to carry the `Int` here, it is linked to the type being handled.

Would be nice to extend this so that clang handles the constant.

Reply:
Good eye! The builtins should be fixed in https://github.com/intel/llvm/pull/6893/commits/ced5db7da5ad56658195e9bc6f5894d811c59f36. 

> It is a bit silly to carry the `Int` here, it is linked to the type being handled.
>
> Would be nice to extend this so that clang handles the constant.

I agree. I am not sure there is much point in having it, and I am not sure we would actually need any special handling from Clang's side. To remove it, I only think we need the SPIR-V translator to not expect it. Little outside the scope of this patch though.

Reply:
> Little outside the scope of this patch though.

Way out-of-scope, it is more me ranting to myself.

The comment is really "it would be nice to do that at some point". Right now the builtin is emitted as the translator expect it, which is what we want. That's also making a stable set to implement for CUDA and HIP.